crates/consensus/src/hardforks/ecotone.rs

```diff
@@ -79,118 +89,86 @@ impl super::Hardforks {
-                input: Self::upgrade_to_calldata(NEW_L1_BLOCK),
-            }
-            .seal_slow(),
-        )
-        .encode_2718(&mut buffer);
-        txs.push(Bytes::from(buffer));
-
-        // Update gas price oracle proxy
-        buffer = Vec::new();
-        OpTxEnvelope::Deposit(
+                input: super::upgrade_to_calldata(Self::NEW_L1_BLOCK),
+            },
             TxDeposit {
                 source_hash: Self::update_gas_price_oracle_source(),
                 from: Address::default(),
-                to: TxKind::Call(GAS_PRICE_ORACLE_DEPLOYER),
+                to: TxKind::Call(Self::GAS_PRICE_ORACLE_DEPLOYER),
                 mint: 0.into(),
                 value: U256::ZERO,
                 gas_limit: 50_000,
                 is_system_transaction: false,
-                input: Self::upgrade_to_calldata(GAS_PRICE_ORACLE),
-            }
-            .seal_slow(),
-        )
-        .encode_2718(&mut buffer);
-        txs.push(Bytes::from(buffer));
-
-        // Enable ecotone
-        buffer = Vec::new();
-        OpTxEnvelope::Deposit(
+                input: super::upgrade_to_calldata(Self::GAS_PRICE_ORACLE),
+            },
             TxDeposit {
                 source_hash: Self::enable_ecotone_source(),
-                from: L1_BLOCK_DEPLOYER,
-                to: TxKind::Call(GAS_PRICE_ORACLE),
+                from: Self::L1_BLOCK_DEPLOYER,
+                to: TxKind::Call(Self::GAS_PRICE_ORACLE),
                 mint: 0.into(),
                 value: U256::ZERO,
                 gas_limit: 80_000,
                 is_system_transaction: false,
                 input: Self::ENABLE_ECOTONE_INPUT.into(),
-            }
-            .seal_slow(),
-        )
-        .encode_2718(&mut buffer);
-        txs.push(Bytes::from(buffer));
-
-        // Deploy EIP-4788
-        buffer = Vec::new();
-        OpTxEnvelope::Deposit(
+            },
             TxDeposit {
                 source_hash: Self::beacon_roots_source(),
-                from: EIP4788_FROM,
+                from: Self::EIP4788_FROM,
                 to: TxKind::Create,
                 mint: 0.into(),
                 value: U256::ZERO,
                 gas_limit: 250_000,
                 is_system_transaction: false,
                 input: Self::eip4788_creation_data(),
-            }
-            .seal_slow(),
-        )
-        .encode_2718(&mut buffer);
-        txs.push(Bytes::from(buffer));
+            },
+        ])
+        .into_iter()
+    }
+
+    /// Returns the list of [OpTxEnvelope]s for the Ecotone network upgrade.
+    pub fn enveloped_txs() -> impl Iterator<Item = OpTxEnvelope> {
+        Self::deposits().map(|deposit| OpTxEnvelope::from(deposit.seal_slow()))
+    }
+}
 
-        txs
+impl super::Hardfork for Ecotone {
+    /// Constructs the Ecotone network upgrade transactions.
+    fn txs(&self) -> impl Iterator<Item = Bytes> + '_ {
+        Self::enveloped_txs().map(|tx| {
```

Comment:
this should probably use deposits direclty?

Reply:
We need to encode the `OpTxEnvelope` wrapper type which internally hashes `TxDeposit` in it's [`From<TxDeposit> for OpTxEnvelope`](https://github.com/alloy-rs/op-alloy/blob/main/crates/consensus/src/transaction/envelope.rs#L71-L75) impl.

Reply:
I've added tests to assert the bytes. If you try to use `TxDeposit` itself, with rlp encoding, it doesn't match the desired bytes.

Reply:
Let me know if I'm misunderstanding but I think you meant using the `deposits` method directly e.g.

```rust
https://github.com/alloy-rs/op-alloy/blob/main/crates/consensus/src/transaction/envelope.rs#L71-L75
```

Reply:
this is the wrong encode fn, you need 2718?

see how encodable2718 is impl for optxenvelope

Reply:
Ah I missed the 2718 methods. Great! I removed and we can follow up with another pr to move 2718 methods into the `Encodable2718/Decodable2718` trait impls.